src/encoded/audit/file.py

```diff
@@ -373,6 +373,41 @@ def audit_file_in_correct_bucket(value, system):
         )
 
 
+def audit_file_matching_md5sum(value, system):
+    '''
+    Files with md5 sums matching other files should be marked with a WARNING audit.
+    If the other files are listed as matching but in fact have different md5 sums,
+    the file should be flagged with an ERROR for incorrect metadata.
+    '''
+
+    matching_files = []
+    checked_statuses = ['released', 'revoked', 'archived', 'in progress']
+    if 'matching_md5sum' not in value or value.get('status') not in checked_statuses:
+        return
+
+    for file in value.get('matching_md5sum'):
+        if file.get('md5sum') != value.get('md5sum'):
+            detail = 'File {} is listed as having a matching md5 sum '.format(file['@id']) + \
+                     'as file {}, but the files have different md5 sums.'.format(value['@id'])
+            yield AuditFailure('Incorrect matching_md5sum', detail, level='ERROR')
+        else:
+            matching_files.append(file['@id'])
+
+    if len(matching_files) > 2:
+        matching_files_joined = '{}, and {}'.format(
+            ', '.join(matching_files[:-1]),
+            matching_files[-1:][0]
+        )
+    else:
+        matching_files_joined = ' and '.join(matching_files)
+
+    detail = 'The md5 sum of file {} '.format(value['@id']) + \
+             'matches that of file(s) {}.'.format(matching_files_joined)
+    yield AuditFailure('Matching md5 sums', detail, level='WARNING')
```

Comment:
Should this be optional if `matching_files` is empty?

Reply:
I'll add a case for len(matching_files)= 0.